lib/src/runner/reporter/xunit.dart

```diff
@@ -0,0 +1,353 @@
+  }
+
+  /// Calculate the test totals for a suite and its children
+  XunitTestSuite _suiteResults(Map group, XunitTestSuite suite) {
+    Map groupStructure = group;
+    XunitTestSuite currentSuite = suite;
+
+    groupStructure.forEach((key, value) {
+      if (key == 'testResults') {
+        value.testResults.forEach((element) {
+          currentSuite.tests++;
+          if (element.skipped) {
+            currentSuite.skipped++;
+          }
+          if (element.error.isNotEmpty) {
+            if (element.error['failure'] == true) {
+              currentSuite.failed++;
+            } else {
+              currentSuite.errored++;
+            }
+          }
+        });
+      } else {
+        _suiteResults(groupStructure[key], currentSuite);
+      }
+    });
+
+    return suite;
+  }
+
+  /// Format testsuite headings
+  String _formatTestSuiteHeading(Map group) {
+    XunitTestSuite suite = _suiteResults(group, group['testResults']);
+
+    String suiteHeading = '';
+    if (suite.tests > 0) {
+      suiteHeading += 'tests="${suite.tests}" ';
+    }
+    if (suite.failed > 0) {
+      suiteHeading += 'failures="${suite.failed}" ';
+    }
+    if (suite.errored > 0) {
+      suiteHeading += 'errors="${suite.errored}" ';
+    }
+    if (suite.skipped > 0) {
+      suiteHeading += 'skipped="${suite.skipped}" ';
+    }
+    return suiteHeading.trimRight();
+  }
+
+  /// A method used to create a nested xml hierarchy
+  String _formatXmlHierarchy(Map xmlMap, {int counter: 0}) {
+    counter++;
+    String result = "";
+    xmlMap.keys.forEach((elem) {
+      if (xmlMap[elem] is Map) {
+        var space = ' ';
+        result += space * counter +
+            '<testsuite name="$elem" ${_formatTestSuiteHeading(xmlMap[elem])}>${_formatTestResults(xmlMap[elem]["testResults"].testResults, counter)}\n';
+        result += _formatXmlHierarchy(xmlMap[elem], counter: counter);
+        result += space * counter + '</testsuite>';
+        if (counter > 0) {
+          result += '\n';
+        }
+      }
+    });
+    return result;
+  }
+
+  /// A callback called when the engine is finished running tests.
+  ///
+  /// [success] will be `true` if all tests passed, `false` if some tests
+  /// failed, and `null` if the engine was closed prematurely.
+  void _onDone(bool success) {
+    // A null success value indicates that the engine was closed before the
+    // tests finished running, probably because of a signal from the user, in
+    // which case we shouldn't print summary information.
+    if (success == null) return;
+
+    print('<?xml version="1.0" encoding="UTF-8" ?>');
+    print(
+        '<testsuite name="All tests" tests="${_engine.passed.length + _engine.skipped.length +_engine.failed.length}" '
+        'errors="$_errorCount" failures="$_failureCount" skipped="${_engine.skipped.length}">');
+
+    if (_groupStructure['testResults']?.testResults == null) {
+      print(_formatXmlHierarchy(_groupStructure).trimRight());
```

Comment:
if I don't include this an extra `\n` is included based on the formatting that I'm doing in `_formatTestResults`